_posts/2023-11-17-transactional-outbox.md

```diff
@@ -0,0 +1,295 @@
+4. **Event Deletion/Mark as Processed:** Once the event is successfully published to the message bus, it is either deleted from the outbox or marked as processed, preventing it from being published again.
+
+By using the transactional outbox, the application ensures that the state in the database and the published events are always consistent.
+Moreover, the pattern exhibits a remarkable resilience to system failures. If the application crashes after committing the transaction but before the event is published, the separate process ensures that the event will still be published later.
+Another significant advantage of the Transactional Outbox Pattern is the decoupling it offers. By separating the concerns of database transactions and event publishing, the pattern allows each to operate independently. This separation not only simplifies the architecture but also enhances scalability and maintainability. 
+It allows the system to evolve more flexibly, adapting to changes and scaling components independently without disrupting the fundamental operations of transaction management and event distribution.
+
+Additionally, an often overlooked yet significant aspect of the Transactional Outbox Pattern is its efficiency, 
+particularly regarding the speed of writing to the outbox. 
+Writing an event to the outbox involves merely appending a row to a table, which is one of the fastest operations in database management. This efficiency means that the pattern does not impose a notable performance penalty.
+
+# Challenges and Solutions
+
+Implementing the Transactional Outbox Pattern, while beneficial, brings its own set of challenges that need careful consideration and handling.
+
+## At least once delivery
+
+When implementing the Transactional Outbox Pattern in a distributed system, it's important to understand that this pattern inherently guarantees at-least-once-delivery semantics. 
+At-least-once-delivery is a delivery approach where the system ensures that a message is definitely delivered to its destination at least once.
+Since the delivery mechanism is designed to ensure that messages are not lost, it errs on the side of sending duplicates rather than risking a message not being delivered at all. 
+As a result, the receiving services or components must be equipped to identify and appropriately handle these duplicate messages. 
+This involves implementing logic to detect if a message has already been processed and ensuring that processing it again does not adversely affect the system's state or lead to incorrect operations.
+
+A common approach to handle duplicate messages is implementing idempotency at the technical level. 
+This involves recording the IDs of consumed events and checking against this record before processing new events. 
+If an incoming event's ID matches one that has already been processed, it is ignored, thus preventing duplicate processing.
+
```

Comment:
how do I know that two databases with their own outboxes are not generating the same id for their separate events? just curious, probably not interesting for the blog Reader

Reply:
In distributed systems, UUIDs are used. The mathematical probability of two UUIDs clashing is extremely low, often likened to the probability of finding a specific grain of sand on a beach, or even in the galaxy. This extremely low probability is what makes them reliable for use in distributed systems.